numpyro/contrib/module.py

```diff
@@ -49,12 +49,16 @@ def flax_module(name, nn_module, *, input_shape=None, **kwargs):
         args = (jnp.ones(input_shape),) if input_shape is not None else ()
         # feed in dummy data to init params
         rng_key = numpyro.prng_key()
-        _, nn_params = nn_module.init(rng_key, *args, **kwargs)
+        nn_params = nn_module.init(rng_key, *args, **kwargs)
+
+        # make a mutable copy
+        nn_params = flax.core.unfreeze(nn_params)
+
         # make sure that nn_params keep the same order after unflatten
         params_flat, tree_def = tree_flatten(nn_params)
```

Comment:
I guess we only need to extract `params` field for optimizer:
```python
    params_flat, tree_def = tree_flatten(nn_params["params"])
    ...
return partial(nn_module.apply, {"params": nn_params})
```
This will keep the params ingestion the same as before (i.e. no prefix `params`). However, this will throw away important mutable variables like `batch_stats` in [nn.BatchNorm](https://flax.readthedocs.io/en/latest/notebooks/flax_basics.html#Variables-and-collections-of-variables).

I'm not sure what's a good solution here. Probably it is better to stick with the old pattern (of setting priors) and think about how to address extra mutable stats in another PR. What do you think?

Reply:
One solution to address extra mutable variables is to define inside this `flax_module` primitive something like this
```python
def apply_fn(*args, **kwargs_)
    variables = {"params": nn_params}
    mutable_vars = kwargs.get("mutable", [])
    if mutable_vars:
        for var in mutable_vars:
            # get the current mutable value
            mutable = numpyro.param_like(name + "$" + var)
            variables[var] = mutable  # jax.lax.stop_gradient(mutable)
        out, variables = nn_module.apply(variables, *args, **kwargs_)
        for var, mutable in variables.items():
            # update the current mutable value
            numpyro.param_like(name + "$" + var, mutable)
    else:
        out = nn_module.apply(variables, *args, **kwargs_)
    return out

return apply_fn
```
Here `SVI` will recognize those variables defined in `param_like` primitives and store it in `SVIState`.